assets/src/store.rs

```diff
@@ -0,0 +1,95 @@
+use std::error::Error;
+use std::fs::File;
+use std::io::Error as IoError;
+use std::path::PathBuf;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+/// An `Allocator`, holding a counter for producing unique ideas for the stores.
+#[derive(Debug, Default)]
+pub struct Allocator {
+    store_count: AtomicUsize,
+}
+
+impl Allocator {
+    /// Creates a new `Allocator`.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Produces a new store id.
+    pub fn next_store_id(&self) -> StoreId {
+        StoreId(self.store_count.fetch_add(1, Ordering::Relaxed))
+    }
+}
+
+/// Directory store.
+///
+/// Please note that there is a default directory storage
+/// inside the `Loader`, which is automatically used when you call
+/// `load`. In case you want another, second, directory for assets,
+/// you can instantiate one yourself, too. Please use `Loader::load_from`
+/// then.
+#[derive(Debug)]
+pub struct Directory {
+    id: StoreId,
+    loc: PathBuf,
+}
+
+impl Directory {
+    /// Creates a new directory storage.
+    pub fn new(alloc: &Allocator, loc: PathBuf) -> Self {
+        Directory {
+            id: alloc.next_store_id(),
+            loc: loc,
+        }
+    }
+}
+
+impl Store for Directory {
+    type Error = IoError;
+    type Location = PathBuf;
+
+    fn store_id(&self) -> StoreId {
+        self.id
+    }
+
+    fn load(&self, category: &str, name: &str, ext: &str) -> Result<Vec<u8>, IoError> {
+        use std::io::Read;
+
+        let mut path: PathBuf = self.loc.clone();
+        path.push(category);
+        path.push(name);
+        path.set_extension(ext);
+
+        let mut v = Vec::new();
+        File::open(&path)?.read_to_end(&mut v)?;
+
+        Ok(v)
+    }
+}
+
+/// A trait for asset stores, which provides
+/// methods for loading
+pub trait Store {
+    type Error: Error + 'static;
+    type Location;
+
+    /// Loads the bytes given a category, id and extension of the asset.
+    fn load(&self, category: &str, id: &str, ext: &str) -> Result<Vec<u8>, Self::Error>;
+
+    /// Returns the unique store id. You'll often want to just
+    /// have such a field for your storage which is initialized using
+    /// `Allocator::next_store_id`.
+    fn store_id(&self) -> StoreId;
+}
+
+/// A unique store id, used to identify the storage in `AssetSpec`.
+#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct StoreId(usize);
+
+impl StoreId {
+    /// Returns a copy of the internal id.
+    pub fn id(&self) -> usize {
```

Comment:
`Deref` and `DerefMut` from all the discussioms Ive heard about it, are mainly for wrapper types. So that does include smart pointers, but isnt limited to.